common/app-starters-function-common/src/main/java/org/springframework/cloud/stream/app/function/StreamAppFunctionAutoConfiguration.java

```diff
@@ -0,0 +1,86 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.cloud.stream.app.function;
+
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.boot.autoconfigure.condition.ConditionalOnBean;
+import org.springframework.cloud.stream.function.IntegrationFlowFunctionSupport;
+import org.springframework.cloud.stream.messaging.Source;
+import org.springframework.context.annotation.Bean;
+import org.springframework.context.annotation.Configuration;
+import org.springframework.integration.channel.AbstractMessageChannel;
+import org.springframework.integration.channel.FluxMessageChannel;
+import org.springframework.integration.channel.QueueChannel;
+import org.springframework.integration.dsl.IntegrationFlow;
+import org.springframework.integration.dsl.IntegrationFlowBuilder;
+import org.springframework.integration.dsl.IntegrationFlows;
+import org.springframework.messaging.Message;
+import org.springframework.messaging.MessageChannel;
+import org.springframework.messaging.PollableChannel;
+import org.springframework.messaging.support.ChannelInterceptor;
+
+/**
+ * Auto configuration class that provides necessary beans and configuration for adding
+ * Function composition support to the stream apps.
+ *
+ * @author Ilayaperumal Gopinathan
+ */
+@Configuration
+@ConditionalOnBean(IntegrationFlowFunctionSupport.class)
+public class StreamAppFunctionAutoConfiguration {
+
+	@Autowired(required = false)
+	private Source channel;
+
+	@Autowired
+	private IntegrationFlowFunctionSupport functionSupport;
+
+	@SuppressWarnings("resource")
+	@Bean
+	public MessageChannel processOutput() {
+		return new FluxMessageChannel();
+	}
+
+	@Bean
+	public IntegrationFlow httpProcessOutput() {
+		MessageChannel processOutput = processOutput();
+		//todo: check multiple function names and add interceptor for each function
+		//todo: this in fact needs to come from SCSt
+		QueueChannel queueChannel  = new QueueChannel();((AbstractMessageChannel) this.channel.output()).addInterceptor(new FunctionInvokerInterceptor(processOutput, queueChannel));
+		IntegrationFlowBuilder flowBuilder = IntegrationFlows.from(processOutput).bridge();
+		this.functionSupport.andThenFunction(flowBuilder, queueChannel);
+		return flowBuilder.get();
+	}
+
+	public class FunctionInvokerInterceptor implements ChannelInterceptor {
+
+		private final MessageChannel processOutput;
+
+		private final PollableChannel pollableChannel;
+
+		public FunctionInvokerInterceptor(MessageChannel processOutput, PollableChannel pollableChannel) {
+			this.processOutput = processOutput;
+			this.pollableChannel = pollableChannel;
+		}
+
+		@Override
+		public Message<?> preSend(Message<?> message, MessageChannel channel) {
+			this.processOutput.send(message);
+			return pollableChannel.receive();
```

Comment:
This is something wrong and fully abusing for `ChannelInterceptor` abstraction.
More over you break here a purpose of the integration flow and hide too much things under a single `ChannelInterceptor`.

I understand what you are pursuing with this `this.channel.output()` "mocking", but that's not what would be great to show for end-users to mock Spring Integration in the end at all. 😉

Reply:
Yes, the above is wrong and @ilayaperumalg and I have discussed it. The interceptor strategy will not work here. There is a better/cleaner and fully reactive alternative.